### PR TITLE
Support for get_keyword_source

### LINKED ISSUE #26: Add support for get_keyword_source API method for Robot Framework 3.2
Robot Framework 3.2 will enhance the dynamic library API and add to`get_keyword_source` API call. See more details in https://github.com/robotframework/robotframework/issues/3507 and  https://github.com/robotframework/robotframework/issues/3516 Add support to PythonLibCore for the new API.

### PR DESCRIPTION
**Pull request by aaltat**
Fixes #26

### CODE REVIEW

File: src/robotlibcore.py [hunk 1]
```diff
@@ -174,6 +175,34 @@ def __join_defaults_with_types(self, method, types):
...
+        method = self.__get_keyword(keyword_name)
+        path = self.__get_keyword_path(method)
+        line_number = self.__get_keyword_line(method)
+        if not path and not line_number:
+            return None
+        if not path:
+            return ':%s' % line_number
+        if not line_number:
+            return path
+        return '%s:%s' % (path, line_number)
```
Comment (pekkaklarck) [inline src/robotlibcore.py]: Might be a bit easier to read without `not`s:

```python
if path and lineno:
    return ...
if path:
    return ...
if lineno:
    return ...
return None

Comment (pekkaklarck) [inline src/robotlibcore.py]: These `if`s might be a bit easier to read without `not`s:

```python
if path and lineno:
    return ...
if path:
    return ...
if lineno:
    return ...
return None

File: src/robotlibcore.py [hunk 2]
```diff
@@ -174,6 +175,34 @@ def __join_defaults_with_types(self, method, types):
...
+
+    def __get_keyword_line(self, method):
+        try:
+            source, line_number = inspect.getsourcelines(method)
+        except (OSError, IOError):
```
Comment (pekkaklarck) [inline src/robotlibcore.py]: I think this can also fail with `TypeError`.

File: src/robotlibcore.py [hunk 3]
```diff
@@ -174,6 +175,34 @@ def __join_defaults_with_types(self, method, types):
...
+            return None
+        for line in source:
+            if 'def' in line:
```
Comment (pekkaklarck) [inline src/robotlibcore.py]: This is a bit risky as `def` can at least in theory be part of a decorator name or argument. If we want to find the real `def` line, `if line.strip().startswith('def ')` would be a bit better.

File: src/robotlibcore.py (hunk 2 above)
Comment (aaltat) [reply src/robotlibcore.py]: Could be, added.

File: src/robotlibcore.py [hunk 4]
```diff
@@ -174,6 +175,34 @@ def __join_defaults_with_types(self, method, types):
...
+
+    def __get_keyword_line(self, method):
+        try:
+            source, line_number = inspect.getsourcelines(method)
+        except (OSError, IOError, TypeError):
+            return None
+        for line in source:
+            if line.strip().startswith('def'):
+                return line_number
+            line_number += 1
```
Comment (pekkaklarck) [inline src/robotlibcore.py]: Hmm, this doesn't work correctly if someone has a library like this:

```python
class LambdaLib:
    kw = lambda self: 42
```

Above is obviously rather strange but from RF point of view it's valid. I see two solutions:

- Return the original `line_number` if `def` line isn't found at all.
- Just return whatever `getsourcelines` returns.

Comment (pekkaklarck) [reply src/robotlibcore.py]: A bigger problem is that decorators returning different function than the original aren't supported correctly. Same issue on Robot side that the moment:
https://github.com/robotframework/robotframework/issues/3516#issuecomment-605635675

**Pull request merged**